### PR TITLE
Support editing username + avatar image [QC-180]

### PR DESCRIPTION
**Pull request by dan13ram**
- "edit profile" button in profile page for logged in users
- can now access profile with `profile/<address>` | `profile/<ens>` | `profile/<username>`
- displaying users' uploading avatar image instead of generated image when available

- [9928df9](https://github.com/quest-chains/dapp/pull/175/commits/9928df9f6e3545e13b320f9e7a9f740c5ad978c4) removes alchemy related errors!!! (cc: @vidvidvid)
- [a1efc21](https://github.com/quest-chains/dapp/pull/175/commits/a1efc21087346b5477fa0938af5333d6584c34f6) closes https://github.com/quest-chains/dapp/issues/162

### CODE REVIEW

**Comment by vidvidvid** [comment]
@dan13ram maybe just make the required length of usernames greater or equal to 3, please :D

File: hooks/usePoH.ts [hunk 1]
```diff
@@ -29,16 +32,18 @@ export const usePoH = (
...
+      setFetching(false);
     }
   }, [address]);
 
   useEffect(() => {
-    getPOH();
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [address]);
+    getPoH();
+  }, [getPoH]);
```
Comment (vidvidvid) [inline hooks/usePoH.ts:42]: when i put gePoH into the dependencies array, the hook kept rerendering the quest chain page, so all the actions were blocked and the tab froze. i went to look if it works, but noticed that there's another error that's popping - `GraphQLError: indexing_error at rehydrateGraphQlError`
could you leave the dependency as it was before, to not freeze the quest chain page if the query starts working again?

**Review by vidvidvid — CHANGES_REQUESTED** [review]

File: components/ProfileView/EditProfileModal.tsx [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  const handleUpdate = useCallback(async () => {
+    let tid;
+    try {
+      setLoading(true);
+      const username = usernameRef.current;
+      if (!username) {
+        setUsernameError('Username cannot be empty.');
+        return;
+      }
+      if (username.length < 4) {
```
[the remark this answers is not in the document]
Comment (dan13ram) [reply components/ProfileView/EditProfileModal.tsx]: fixed in [f7ce284](https://github.com/quest-chains/dapp/pull/175/commits/f7ce284c2c239ad87d23a1352f8008a4fe0d1d34)

File: hooks/usePoH.ts (hunk 1 above)
Comment (dan13ram) [reply hooks/usePoH.ts:42]: @vidvidvid this hook wasn't being used in the quest chain page but `usePoHs` was.. So i took a look at the code and added a fix in [26a0c80](https://github.com/quest-chains/dapp/pull/175/commits/26a0c80b42012b729eca9c49d39933d9ce102096)

File: hooks/usePoH.ts [hunk 3]
```diff
@@ -54,7 +54,8 @@ export const usePoHs = (roles: {
     [addr: string]: boolean;
   };
 } => {
-  const addresses = Object.keys(roles);
+  const addresses = useMemo(() => Object.keys(roles), [roles]);
```
Comment (dan13ram) [inline hooks/usePoH.ts:57]: @vidvidvid this is the main change that fixed the infinite re-render

Comment (vidvidvid) [reply hooks/usePoH.ts:57]: aaaah, nice one. thank you for fixing it! :)

**Comment by vidvidvid** [comment]
@dan13ram after setting my name to vid & uploading the pfp, the app navigated to `/`. i think it tried to navigate to my new profile url, but encountered an error, which appeared as i tried to navigate to `/profile/vid` manually:
```
GET https://deploy-preview-175--quest-chains-dev.netlify.app/_next/data/A3ChEZ1pFmaxYoic5QUS_/profile/vid.json?name=vid 404
```

**Comment by dan13ram** [comment]
@vidvidvid My bad. I hadn't changed all places for 3 length usernames so it was not finding your profile 😓 
Its fixed now in [ef7f861](https://github.com/quest-chains/dapp/pull/175/commits/ef7f861a0133cb1ba983b7066355f14b42b3167e)

Can you please test again ?

**Comment by vidvidvid** [comment]
@dan13ram all good now!
i'm REALLY happy about the disappearance of the alchemy error :') the gravatar package was causing it?

**Comment by dan13ram** [comment]
> the gravatar package was causing it?

Yes that was it @vidvidvid. Do we merge ?

**Review by vidvidvid — APPROVED** [review]
yep, let's merge!

**Pull request merged**